tsdb/index/index.go

```diff
@@ -425,65 +452,124 @@ func (w *Writer) WriteLabelIndex(names []string, values []string) error {
 		offset: w.pos,
 	})
 
-	w.buf2.Reset()
-	w.buf2.PutBE32int(len(names))
-	w.buf2.PutBE32int(valt.Len())
+	startPos := w.pos
+	if err := w.write([]byte("alen")); err != nil {
+		return err
+	}
+	w.crc32.Reset()
+
+	w.buf1.Reset()
```

Comment:
The same flow is repeating 5 times. I guess it's hard to generalize due to different formats of each section, but wonder if something like this would help?

```
func (w *Writer) writeSection(func init(*EncBuf), func next(*EncBuf) bool) error
```
... ideally with `*EncBuf` being under some interface for `Put*` methods only. Or it's overkill?